docs/tasks/administer-cluster/kubeadm-upgrade-ha.md

```diff
@@ -0,0 +1,176 @@
+
+First, the master nodes are upgraded (run `kubectl cordon` / `kubectl uncordon` from whatever machine you have set up to issue `kubectl` commands):
+
+```shell
+kubectl cordon <MASTER-NODE-FQDN>
+systemctl stop keepalived kubelet docker
+# use your distro's package manager, e.g. 'yum' on RH-based systems
+# for the versions stick to kubeadm's output (see above)
+yum install -y kubelet-<NEW-K8S-VERSION> kubectl-<NEW-K8S-VERSION> kubeadm-<NEW-K8S-VERSION> kubernetes-cni-<NEW-CNI-VERSION>
+systemctl start `echo keepalived kubelet docker | tac`
+kubectl uncordon <MASTER-NODE-FQDN>
+```
+
+In this example an _rpm_-based system is assumed and `yum` is used for installing the upgraded software. On _deb_-based systems it will be `apt-get update` and then `apt-get install <PACKAGE>=<NEW-K8S-VERSION>` for all packages.
+
+On the worker nodes there is no keepalived to stop and start and all processes need to be moved to other nodes cleanly before shutting anything down, hence `drain` is used instead of `cordon`. In this process all containers will get moved from one node to another at least once. If some of them keep local data you will need some strategy for this scenario. This really depends on the application(s) running in the cluster, and there is no simple, generic solution to this.
+
+On the master (or wherever you run `kubectl` commands):
+
+```shell
+kubectl drain --delete-local-data --ignore-daemonsets <WORKER-NODE-FQDN>
+```
+
+On the respective worker node:
+
+```shell
+systemctl stop keepalived kubelet docker
+# use your distro's package manager, e.g. 'yum' on RH-based systems
+# for the versions stick to kubeadm's output (see above)
+yum install -y kubelet-<NEW-K8S-VERSION> kubectl-<NEW-K8S-VERSION> kubeadm-<NEW-K8S-VERSION> kubernetes-cni-<NEW-CNI-VERSION>
+systemctl start `echo keepalived kubelet docker | tac`
+```
+
+Like above, for _deb_-based systems use `apt-get install <PACKAGE>=version` to upgrade the packages to the respective desired versions.
+
+Now the new version of the `kubelet` should be running on the host. Verify this using the following command on the respective host:
+
+```shell
+systemctl status kubelet
+```
+
+If everything was successful, again on the master (or wherever you run `kubectl` commands):
+
+```shell
+kubectl uncordon <WORKER-NODE-FQDN>
+```
+
+Verify that the upgraded node is available again by executing the following from wherever you run `kubectl` commands:
+
+```shell
+kubectl get nodes
+```
+
+If the `STATUS` column of the above command shows `Ready` for the upgraded host, you can continue (you may have to repeat this for a couple of time before the node gets `Ready`).
```

Comment:
Maybe this answers the previous question I had regarding a better way to know when to move on to secondary masters? If so, I guess that previous sentence could be removed.

Reply:
No, I always wait until nodes are listed as "ready", but I still run into the failures described above.